sdk/servicebus/service-bus/src/core/messageSender.ts

```diff
@@ -419,6 +414,8 @@ export class MessageSender extends LinkEntity {
           );
           this.isConnecting = true;
           await this._negotiateClaim();
+          checkAborted();
```

Comment:
`waitForTimeoutOrAbortOrResolve` will ensure that it throws the `AbortError` when the signal is fired.
Meaning, when we call `waitForTimeoutOrAbortOrResolve` from inside send(), the promise is rejected appropriately due to the AbortError thrown by `waitForTimeoutOrAbortOrResolve`.

In such cases, what happens to the error thrown at this point?

Can we make it so that the caller of `open()` is responsible to throw the AbortError, and `open()` is only responsible for exiting as soon as possible?

Reply:
We might be overly covering abortSignal now :)

I do think the way open() is checking is correct though - early exiting would be nice but early exiting on error seems a bit too chummy with the caller. With that said, I think you're saying that we might lose a legitimate error somehow but I'm not following the sequence to get there.

Reply:
> but early exiting on error seems a bit too chummy with the caller.

I agree for normal errors, but I feel that the `AbortError` and the timeout error is of our own making and that it would be simpler to handle them from 1 place. We don't really check if the timer has run out from each place inside open() and throw the error, then why do we feel the need to do it for `AbortError`?

> With that said, I think you're saying that we might lose a legitimate error somehow but I'm not following the sequence to get there.

I am concerned with a `unhandledpromiserejectionwarning` scenario where errors are thrown, but there is no nice way for them to bubble up so that the user can handle them.

Reply:
> 
> 
> > but early exiting on error seems a bit too chummy with the caller.
> 
> I agree for normal errors, but I feel that the `AbortError` and the timeout error is of our own making and that it would be simpler to handle them from 1 place. We don't really check if the timer has run out from each place inside open() and throw the error, then why do we feel the need to do it for `AbortError`?

The big difference between the abortSignal and the timeout is that the timeout is completely self-imposed and managed by our code. So we can completely manage it's lifetime and it's completely self-contained. 

The abortSignal isn't triggered by us at all, so we have to check it since we wouldn't have any other way of knowing - the triggering and unrolling of the abort signal is very much a cooperative activity. 

From an "own making" standpoint I think I get it but it seems like a nice goal to say "user errors and service errors are routed and handled the same" which is basically what I'm doing here.

> 
> > With that said, I think you're saying that we might lose a legitimate error somehow but I'm not following the sequence to get there.
> 
> I am concerned with a `unhandledpromiserejectionwarning` scenario where errors are thrown, but there is no nice way for them to bubble up so that the user can handle them.

Ah. I think we're actually pretty well covered with the way it's written - each function is doing the "right" thing with regards to error propagation from what I can tell (and what I have in my tests). I don't think we're swallowing or ignoring errors. For the most part most of the code is just ensuring that some `finally` blocks run but the end goal still gets delivered to the user.

Reply:
I tested the case when the function passed to `waitForTimeoutOrAbortOrResolve` throws an error after the abort signal is fired. There is no `unhandledpromiserejectionwarning` and the error gets swallowed up. 

```js
async function open() {
  throw new Error("boo! Do you see me?");
}

async function main() {
  const controller = new AbortController();
  const signal = controller.signal;
  controller.abort();
  
  await waitForTimeoutOrAbortOrResolve({
    actionFn: open,
    abortMessage: "This is the abort message",
    abortSignal: signal,
    timeoutMs: 1000,
    timeoutMessage: "This is the timeout message"
  })
}
main().catch(err => {
  console.log("I caught the error!!", err);
})
```

Reply:
I tested the case when the function passed to `waitForTimeoutOrAbortOrResolve` throws an error after the abort signal is fired. There is no `unhandledpromiserejectionwarning`, but the error does get swallowed up which should be fine I guess.

```js
async function open() {
  throw new Error("boo! Do you see me?");
}

async function main() {
  const controller = new AbortController();
  const signal = controller.signal;
  controller.abort();
  
  await waitForTimeoutOrAbortOrResolve({
    actionFn: open,
    abortMessage: "This is the abort message",
    abortSignal: signal,
    timeoutMs: 1000,
    timeoutMessage: "This is the timeout message"
  })
}
main().catch(err => {
  console.log("I caught the error!!", err);
})
```